src/main/java/io/strimzi/kafka/quotas/VolumeSource.java

```diff
@@ -58,64 +61,72 @@ public VolumeSource(Admin admin, VolumeObserver volumeObserver, int timeout, Tim
-                    log.debug("Successfully described cluster: " + nodes);
-                }
-                //Deliberately stay on the adminClient thread as the next thing we do is another admin API call
-                onDescribeClusterSuccess(nodes, logDirsPerBrokerPromise);
-            }
-        });
-
         try {
+            log.info("Updating cluster volume usage.");
+            CompletableFuture<VolumeUsageResult> volumeUsageResultPromise = new CompletableFuture<>();
+            log.debug("Attempting to describe cluster");
+            admin.describeCluster().nodes().whenComplete((nodes, throwable) -> {
+                if (throwable != null) {
+                    log.error("Error while describing cluster", throwable);
+                    volumeUsageResultPromise.complete(failure(VolumeSourceObservationStatus.DESCRIBE_CLUSTER_ERROR, throwable));
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Successfully described cluster: " + nodes);
+                    }
+                    //Deliberately stay on the adminClient thread as the next thing we do is another admin API call
+                    onDescribeClusterSuccess(nodes, volumeUsageResultPromise);
+                }
+            });
             //Bring it back to the original thread to do the actual work of the plug-in.
-            final Map<Integer, Map<String, LogDirDescription>> logDirsPerBroker = logDirsPerBrokerPromise.get(timeout, timeoutUnit);
-            onDescribeLogDirsSuccess(logDirsPerBroker);
+            final VolumeUsageResult result = volumeUsageResultPromise.get(timeout, timeoutUnit);
+            notifyObserver(result);
             log.info("Updated cluster volume usage.");
         } catch (InterruptedException e) {
+            notifyObserver(failure(VolumeSourceObservationStatus.INTERRUPTED, e));
             log.warn("Caught interrupt exception trying to describe cluster and logDirs: {}", e.getMessage(), e);
             Thread.currentThread().interrupt();
-        } catch (ExecutionException | TimeoutException e) {
-            log.warn("Caught exception trying to describe cluster and logDirs: {}", e.getMessage(), e);
+        } catch (ExecutionException e) {
+            volumeObserver.observeVolumeUsage(failure(VolumeSourceObservationStatus.EXECUTION_EXCEPTION, e));
+            log.warn("Caught execution exception trying to describe cluster and logDirs: {}", e.getMessage(), e);
+        } catch (TimeoutException e) {
+            notifyObserver(failure(VolumeSourceObservationStatus.SAFETY_TIMEOUT, e));
+            log.warn("Caught timeout exception trying to describe cluster and logDirs: {}", e.getMessage(), e);
+        } catch (RuntimeException e) {
+            notifyObserver(failure(VolumeSourceObservationStatus.EXCEPTION, e));
+            log.warn("Caught runtime exception trying to describe cluster and logDirs: {}", e.getMessage(), e);
+        }
```

Comment:
Am I wrong or it's not handling the "potential" `IllegalArgumentException` coming from the `failure` method?

Reply:
`IllegalArgumentException`extends `RuntimeException` so will be handled here, were you expecting specific handling of IllegalArgumentException?

Reply:
You are right, no need to change.